### PR TITLE
refactor: Redesign the repair strategies

### LINKED ISSUE #216: Redesign SoraldAbstractRepair to be more testable
There are a number of things that need to be improved here.

* I need a method that returns a Spoon model in order to test some of the finer details
    - It's very hard to test some of the finer details by first writing to a file, as bugs can then be located in the processor that performed the repair, the printer, or the parser
    - This is also a good step toward defining a nice API for Sorald
* The `repair` method does too many things. It reads files, does repairs, write files, and also does user I/O. All of those things need to be pieced out into methods of their own.
* I don't think having `SoraldAbstractRepair` as an abstract base class is meaningful. It would be better of as an interface. The state currently kept in the base class could simply be passed as parameters to the methods. It would go some way to untangling the structure.

### PR DESCRIPTION
**Pull request by slarse**
Fix #216 

This PR is not technically a refactor, as it coincidentally fixes a bug (#170), but it's close enough as to make that the most reasonable label :)

In any case, this PR does away with the strategy pattern that was applied to `SoraldAbstractRepair`, and the `DefaultRepair` and `SegmentRepair` implementations. The abstraction was very leaky, as the two concrete classes functioned fundamentally differently, requiring different options in the config, different handling of processors, etc etc. Code reuse was also poor, and `SegmentRepair` did not handle multiple rules (#170). This redesign fixes the problems in #170 simply by restructuring the code, the segment strategy can now repair multiple rules in the same way that the default does.

The new design is less object-oriented, and in my opinion significantly easier to understand. No inheritance required. This is in large part in preparation of gathering statistics (#193), as well as providing an actual API for Sorald's library use case. Note however that the `Repair` class is _not_ what I consider the public API of Sorald, as I want the public API to not have to deal with the `SoraldConfig` object.

The `repair`, `defaultRepair` and `segmentRepair` methods are however good indications of what the API will eventually look like, and are package-private for testing purposes.

### CODE REVIEW

File: src/main/java/sorald/Repair.java [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+                .map(
+                        segment -> {
+                            processor.initResource(segment, inputBaseDir);
+                            Launcher launcher = createSegmentLauncher(segment);
+                            CtModel model = launcher.getModel();
+                            repairModelWithInitializedProcessor(model, processor);
+                            return model;
+                        })
+                .takeWhile(model -> processor.getNbFixes() < config.getMaxFixesPerRule());
+    }
```
Comment (slarse) [inline src/main/java/sorald/Repair.java:118]: This is essentially what became of the previous inheritance structure. This is much easier for me to wrap my head around. I envision the public API of Sorald to look something like this.

**Comment by slarse** [comment]
@fermadeiral This is ready for review, but it will conflict with #218. So pick whichever one you feel like reviewing first!

**Comment by slarse** [comment]
@fermadeiral Conflict fixed!

**Comment by fermadeiral** [comment]
That's really great! Thanks @slarse!

**Pull request merged**